tests/func/test_add.py

```diff
@@ -662,3 +662,36 @@ def test_not_raises_on_re_add(tmp_dir, dvc):
 
     tmp_dir.gen({"file2": "file2 content", "file": "modified file"})
     dvc.add(["file2", "file"])
+
+
+@pytest.mark.parametrize("link", ["hardlink", "symlink", "copy"])
+def test_add_empty_files(tmp_dir, dvc, link):
+    file = "foo"
+    dvc.cache.local.cache_types = [link]
+    stages = tmp_dir.dvc_gen(file, "")
+
+    assert (tmp_dir / file).exists()
+    assert (tmp_dir / (file + Stage.STAGE_FILE_SUFFIX)).exists()
+    assert os.path.exists(stages[0].outs[0].cache_path)
+
+
+def test_add_optimization_for_hardlink_on_empty_files(tmp_dir, dvc, mocker):
+    dvc.cache.local.cache_types = ["hardlink"]
+    tmp_dir.gen({"foo": "", "bar": "", "lorem": "lorem", "ipsum": "ipsum"})
+    m = mocker.spy(RemoteLOCAL, "is_hardlink")
+    stages = dvc.add(["foo", "bar", "lorem", "ipsum"])
+
+    m.assert_called_once()
```

Comment:
I think spy object does not have this method, you can do `assert m.call_count == 1`

Reply:
Seems like it does: https://github.com/pytest-dev/pytest-mock/#spy

Reply:
@pared, looks like you are right. It fails on Python3.5 for some reasons (on other, it works fine). using what you suggested.

Reply:
`assert_called_once` was only added on Python3.6.  https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once 🤦